src/libraries/System.Private.CoreLib/src/System/Runtime/InteropServices/Marshalling/ReadOnlySpanMarshaller.cs

```diff
@@ -0,0 +1,126 @@
+            {
+                // Emulate the pinning behavior:
+                // If the span is over a null reference, then pass a null pointer.
+                if (Unsafe.IsNullRef(ref MemoryMarshal.GetReference(managed)))
+                {
+                    numElements = 0;
+                    return null;
+                }
+
+                numElements = managed.Length;
+
+                // Always allocate at least one byte when the array is zero-length.
+                int spaceToAllocate = Math.Max(checked(sizeof(TUnmanagedElement) * numElements), 1);
+                return (TUnmanagedElement*)Marshal.AllocCoTaskMem(spaceToAllocate);
+            }
+
+            public static ReadOnlySpan<T> GetManagedValuesSource(ReadOnlySpan<T> managed)
+                => managed;
+
+            public static Span<TUnmanagedElement> GetUnmanagedValuesDestination(TUnmanagedElement* unmanaged, int numElements)
+                => new Span<TUnmanagedElement>(unmanaged, numElements);
+        }
+
+        public ref struct ManagedToUnmanagedIn
+        {
+            // We'll keep the buffer size at a maximum of 200 bytes to avoid overflowing the stack.
+            public static int BufferSize { get; } = 0x200 / sizeof(TUnmanagedElement);
+
+            private ReadOnlySpan<T> _managedArray;
+            private TUnmanagedElement* _allocatedMemory;
+            private Span<TUnmanagedElement> _span;
+
+            /// <summary>
+            /// Initializes the <see cref="SpanMarshaller{T, TUnmanagedElement}.ManagedToUnmanagedIn"/> marshaller.
+            /// </summary>
+            /// <param name="managed">Span to be marshalled.</param>
+            /// <param name="buffer">Buffer that may be used for marshalling.</param>
+            /// <remarks>
+            /// The <paramref name="buffer"/> must not be movable - that is, it should not be
+            /// on the managed heap or it should be pinned.
+            /// </remarks>
+            public void FromManaged(ReadOnlySpan<T> managed, Span<TUnmanagedElement> buffer)
+            {
+                _allocatedMemory = null;
+                // Emulate the pinning behavior:
+                // If the span is over a null reference, then pass a null pointer.
+                if (Unsafe.IsNullRef(ref MemoryMarshal.GetReference(managed)))
+                {
+                    _managedArray = null;
+                    _span = default;
+                    return;
+                }
+
+                _managedArray = managed;
+
+                // Always allocate at least one byte when the span is zero-length.
+                if (managed.Length <= buffer.Length)
+                {
+                    _span = buffer[0..managed.Length];
+                }
+                else
+                {
+                    int bufferSize = checked(managed.Length * sizeof(TUnmanagedElement));
+                    int spaceToAllocate = Math.Max(bufferSize, 1);
```

Comment:
The `NativeMemory.Alloc` API already make the non-null guarantees for passing 0. Is that what this is for?

Reply:
Yes, that's what this is for. I'll change it to just pass `bufferSize`.